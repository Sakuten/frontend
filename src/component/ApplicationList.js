import React from 'react'
import { observer } from 'mobx-react'
import styled from 'styled-components'

const Tag = styled.span.attrs({
  className: 'tag'
})`
  margin: 0.75rem;
`

const ApplicationList = ({list, onCancel}) => (
  <div>
    {
      list ? list.map((c, i) =>
        <div data-test='applicationlist-application' key={i}>
          <div className='card'>
            <header className='card-header'>
              <p className='card-header-title'>
                {c.lottery.name}
              </p>
              {c.status === 'pending' ? <Tag className='is-dark'>発表をお待ちください</Tag> : c.status === 'won' ? <Tag className='is-success'>当選しました。</Tag> : <Tag className='is-danger'>落選しました。</Tag>}
            </header>
            <footer className='card-footer'>
              <a className='card-footer-item has-text-danger' data-test='applicationlist-cancel' onClick={() => onCancel(c.id)}>キャンセル</a>
            </footer>
          </div>
        </div>
      ) : null
    }
  </div>
)

export default observer(ApplicationList)
