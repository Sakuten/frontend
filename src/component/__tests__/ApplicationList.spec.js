import React from 'react'
import { shallow } from 'enzyme'
import ApplicationList from '../ApplicationList'

const setup = propOverrides => {
  const props = Object.assign({
    list: [
      {
        'id': 1,
        'lottery': {
          'classroom_id': 1,
          'done': false,
          'id': 1,
          'index': 0,
          'name': '5A.0',
          'winners': []
        },
        'status': 'pending'
      },
      {
        'id': 2,
        'lottery': {
          'classroom_id': 1,
          'done': false,
          'id': 2,
          'index': 0,
          'name': '5A.0',
          'winners': []
        },
        'status': 'pending'
      }
    ],
    onCancel: jest.fn()
  }, propOverrides)

  const wrapper = shallow(<ApplicationList {...props} />)

  return {
    props,
    wrapper,
    application: wrapper.find('[data-test="applicationlist-application"]'),
    cancelButton: wrapper.find('[data-test="applicationlist-cancel"]'),
    notfound: wrapper.find('[data-test="applicationlist-notfound"]')
  }
}

describe('components', () => {
  describe('ApplicationList', () => {
    it('render', () => {
      const { wrapper } = setup()
      expect(wrapper).toMatchSnapshot()
    })

    it('renders a message when the list was empty', () => {
      const { notfound } = setup({list: []})
      expect(notfound.length).toBe(1)
    })

    it('renders two applications', () => {
      const { application } = setup()
      expect(application.length).toBe(2)
    })

    it('renders two cancel buttons', () => {
      const { cancelButton } = setup()
      expect(cancelButton.length).toBe(2)
    })

    it('calls onCancel with application id when cancel button is clicked', () => {
      const mock = jest.fn()
      const { cancelButton } = setup({list: [{id: 1, lottery: {name: "5A.0"}}], onCancel: mock})
      cancelButton.at(0).simulate('click')
      expect(mock).toBeCalledWith(1)
    })
  })
})
