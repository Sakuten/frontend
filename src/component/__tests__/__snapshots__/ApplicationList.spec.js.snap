// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components ApplicationList render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ApplicationList
    list={
      Array [
        Object {
          "id": 24,
          "lottery_id": 1,
          "status": null,
          "user_id": 3,
        },
        Object {
          "id": 25,
          "lottery_id": 2,
          "status": null,
          "user_id": 3,
        },
      ]
    }
    onCancel={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          data-test="applicationlist-application"
        >
          {"id":24,"lottery_id":1,"status":null,"user_id":3}
          <button
            data-test="applicationlist-cancel"
            onClick={[Function]}
          >
            Cancel
          </button>
        </div>,
        <div
          data-test="applicationlist-application"
        >
          {"id":25,"lottery_id":2,"status":null,"user_id":3}
          <button
            data-test="applicationlist-cancel"
            onClick={[Function]}
          >
            Cancel
          </button>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "{\\"id\\":24,\\"lottery_id\\":1,\\"status\\":null,\\"user_id\\":3}",
            <button
              data-test="applicationlist-cancel"
              onClick={[Function]}
            >
              Cancel
            </button>,
          ],
          "data-test": "applicationlist-application",
        },
        "ref": null,
        "rendered": Array [
          "{\\"id\\":24,\\"lottery_id\\":1,\\"status\\":null,\\"user_id\\":3}",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Cancel",
              "data-test": "applicationlist-cancel",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Cancel",
            "type": "button",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "{\\"id\\":25,\\"lottery_id\\":2,\\"status\\":null,\\"user_id\\":3}",
            <button
              data-test="applicationlist-cancel"
              onClick={[Function]}
            >
              Cancel
            </button>,
          ],
          "data-test": "applicationlist-application",
        },
        "ref": null,
        "rendered": Array [
          "{\\"id\\":25,\\"lottery_id\\":2,\\"status\\":null,\\"user_id\\":3}",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Cancel",
              "data-test": "applicationlist-cancel",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "Cancel",
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            data-test="applicationlist-application"
          >
            {"id":24,"lottery_id":1,"status":null,"user_id":3}
            <button
              data-test="applicationlist-cancel"
              onClick={[Function]}
            >
              Cancel
            </button>
          </div>,
          <div
            data-test="applicationlist-application"
          >
            {"id":25,"lottery_id":2,"status":null,"user_id":3}
            <button
              data-test="applicationlist-cancel"
              onClick={[Function]}
            >
              Cancel
            </button>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "{\\"id\\":24,\\"lottery_id\\":1,\\"status\\":null,\\"user_id\\":3}",
              <button
                data-test="applicationlist-cancel"
                onClick={[Function]}
              >
                Cancel
              </button>,
            ],
            "data-test": "applicationlist-application",
          },
          "ref": null,
          "rendered": Array [
            "{\\"id\\":24,\\"lottery_id\\":1,\\"status\\":null,\\"user_id\\":3}",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Cancel",
                "data-test": "applicationlist-cancel",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Cancel",
              "type": "button",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "{\\"id\\":25,\\"lottery_id\\":2,\\"status\\":null,\\"user_id\\":3}",
              <button
                data-test="applicationlist-cancel"
                onClick={[Function]}
              >
                Cancel
              </button>,
            ],
            "data-test": "applicationlist-application",
          },
          "ref": null,
          "rendered": Array [
            "{\\"id\\":25,\\"lottery_id\\":2,\\"status\\":null,\\"user_id\\":3}",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Cancel",
                "data-test": "applicationlist-cancel",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "Cancel",
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
